
---
title: "ch VDJdb sandbox. ROC curves"
header-includes:
- \usepackage{amsmath,amsthm,amssymb}
- \usepackage{mathtext}
- \usepackage[T1,T2A]{fontenc}
- \usepackage[utf8]{inputenc}
- \usepackage[english,russian]{babel}
output: 
  pdf_document:
    keep_tex: yes
    latex_engine: xelatex
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Here $Q_{k_{i}m_{j}}$ is Q-score for amino acid substitution in aligned sequences $k$ and $m$ at $i$ and $j$  positions respectively. $Q_{k_{i}m_{j}}$ represents $log_{10}$ of odds ratios for probabilities to see substitution of amino acid in same and distinct antigens. lengths of sequences $k$ and $m$ are similar and are equal to $N$. No indels allowed.

To get the weight of an alignment we sum those Q-scores as follows:
$$S_{km} = \sum_{i=1}^N Q_{k_{i}m_{i}}$$

To compute predicted value for sequences $k$ and $m$ to bind same epitope weight of an alignment was divided by it's length $N_{km}$:

$$W_{km} = \frac{S_{km}}{N_{km}}$$

Those values were computed using script "ch_roc_curves.py".

Files with values are in folder "seqdata/HomoSapiens/scores/".
Scripts are in folder "scripts/".

Point to note:
Cysteines were omitted while lengths and weights were calculated—é

Load data from folder and calculate ROC curves:

```{r}
library(pROC)

read_file_for_ROC <- function(folder, sub, del, ins, tot, name) {
  suffix <- paste(sub, del, ins, tot, sep="_")
  inpname <- paste(paste(name, suffix, sep="_"), '.txt', sep='')
  df <- read.csv(paste(folder, inpname, sep =""), sep = '\t', stringsAsFactors = F)
  colnames(df) <- c("sameantigen", "score", "pair")
  df$dataset <- suffix
  return(df)
}

folder <- "seqdata/HomoSapiens/scores/"
par(mfrow=c(2,3))
for (s in 1:6) {
  df <- read_file_for_ROC(folder, s, 0, 0, s, 'pairscores')
  dfroc <- roc(sameantigen ~ score, df)
  print(paste(s,0,0,s,sep='_'))
  print(dfroc)
  print('-----------------')
  plot(dfroc, main=paste(s,0,0,s,sep='_'))
}



```